Automattic/wp-calypso
# Signup: decouple "shouldHideFreePlan" from site creation logic, allow skipping domain with "shouldHideFreePlan" as false

#### Proposed Changes

* After the PR https://github.com/Automattic/wp-calypso/pull/67332 caused a regression, the changes were reverted.
* This PR adds unit tests for the function that generates the payload for new site creation.
* This function was also modified to decouple shouldHideFreePlan from the function. "shouldHideFreePlan" should be used for displaying visual changes to the user, depending whether or not we want to allow the user to skip a purchase, and should not be a part of the payload generator.
* Whenever a user select a domain free or paid, `siteUrl` is provided to the function and is more than enough for us to make the decision whether or not to use autogeneration.

#### Testing Instructions
- Checkout this branch
- Run `yarn start`
- Test on the [newsletter flow](http://calypso.localhost:3000/setup?flow=newsletter)
- And also test on the [link in bio flow](http://calypso.localhost:3000/setup?flow=link-in-bio)
- When clicking on "Decide Later" or selecting the free wordpress domain in the domains page you should see subheader with a link <img width="803" alt="Screenshot 2022-09-02 at 18 08 17" src="https://user-images.githubusercontent.com/7000684/188194851-78a9de14-1234-4898-bbc0-071aacbb6a28.png">
- When selecting a paid domain it should NOT show a link to start with a free plan <img width="662" alt="Screenshot 2022-09-02 at 18 37 50" src="https://user-images.githubusercontent.com/7000684/188199440-f925810d-c325-4f5a-be6c-73952fd12a9b.png">
- For tailored and normal flows test the free domain selection. Select a free wordpress domain and ensure that the domain assigned after the flow finishes is the same as the one you choose.

Related to #67330
